engine.js

```diff
@@ -822,49 +822,58 @@ var accumulateResult = function(ifResult, thenResult) {
      * @param  {array} loanApplicationRegisters An array of the LARs to process
      * @return {array}                          An array of the results
      */
-    HMDAEngine.getTotalsByMSA = function(loanApplicationRegisters) {
-        return Promise.all(_.chain(loanApplicationRegisters)
-        .groupBy('metroArea')
-        .collect(function(value, key) {
-            return this.getMSAName(key).then(function(msaName) {
-                var result = {msaCode: key, msaName: msaName, totalLAR: 0, totalLoanAmount: 0, totalConventional: 0, totalFHA: 0, totalVA: 0, totalFSA: 0,
-                    total1To4Family: 0, totalMFD: 0, totalMultifamily: 0, totalHomePurchase: 0, totalHomeImprovement: 0, totalRefinance: 0};
-
-                _.each(value, function(element) {
-                    result.totalLAR++;
-                    result.totalLoanAmount += +element.loanAmount;
-
-                    if (element.loanType === '1') {
-                        result.totalConventional++;
-                    } else if (element.loanType === '2') {
-                        result.totalFHA++;
-                    } else if (element.loanType === '3') {
-                        result.totalVA++;
-                    } else if (element.loanType === '4') {
-                        result.totalFSA++;
-                    }
+    HMDAEngine.getTotalsByMSA = function(hmdaFile) {
+        // get the msa branch list for depository to reduce calls to API
+        var depository = (hmdaFile.transmittalSheet.agencyCode==='7') ? false : true;
+        return this.getMSABranches(hmdaFile.transmittalSheet.agencyCode, hmdaFile.transmittalSheet.respondentID)
+        .then(function (branchResult) {
+            return Promise.all(_.chain(hmdaFile.loanApplicationRegisters)
+            .groupBy('metroArea')
+            .filter(function (msaCode) {
+                return (!depository && msaCode.length>=5) || 
+                    (depository && _.contains(branchResult,msaCode[0].metroArea.toString())) || 
```

Comment:
Calling `filter()` as above would allow you to do `_.contains(branchResult, metroArea)`